firmware/quadruna/RSM/src/tasks.c

```diff
@@ -45,52 +87,162 @@ AdcChannel id_to_adc[] = {
     HAL_TIM_Base_Start(&htim3);
 
+    hw_hardFaultHandler_init();
+    hw_can_init(&hcan1);
+
+    io_canTx_init(io_jsoncan_pushTxMsgToQueue);
+    io_canTx_enableMode(CAN_MODE_DEFAULT, true);
+    io_can_init(&can_config);
     io_chimera_init(&debug_uart, GpioNetName_rsm_net_name_tag, AdcNetName_rsm_net_name_tag);
 
+    app_canTx_init();
+    app_canRx_init();
+
+    app_globals_init(&config);
+
+    app_heartbeatMonitor_init(
+        heartbeatMonitorChecklist, heartbeatGetters, heartbeatUpdaters, &app_canTx_RSM_Heartbeat_set,
+        heartbeatFaultSetters, heartbeatFaultGetters);
+
+    app_stateMachine_init(app_mainState_get());
+
+    app_canTx_RSM_Hash_set(GIT_COMMIT_HASH);
+    app_canTx_RSM_Clean_set(GIT_COMMIT_CLEAN);
     // TODO: Re-enable watchdog.
 }
 
 void tasks_run100Hz(void)
 {
-    // TODO: Setup tasks.
-    osDelay(osWaitForever);
+    static const TickType_t period_ms   = 10;
+    static uint32_t         start_ticks = 0;
+    start_ticks                         = osKernelGetTickCount();
+
+    for (;;)
+    {
+        const uint32_t start_time_ms = osKernelGetTickCount();
+
+        app_stateMachine_tick100Hz();
+        io_canTx_enqueue100HzMsgs();
+
+        start_ticks += period_ms;
+        osDelayUntil(start_ticks);
+    }
 }
 
 void tasks_runCanTx(void)
 {
-    // TODO: Setup tasks.
-    osDelay(osWaitForever);
+    for (;;)
+    {
+        io_can_transmitMsgFromQueue();
+    }
 }
 
 void tasks_runCanRx(void)
 {
-    // TODO: Setup tasks.
-    osDelay(osWaitForever);
+    for (;;)
+    {
+        CanMsg rx_msg;
+        io_can_popRxMsgFromQueue(&rx_msg);
+
+        JsonCanMsg jsoncan_rx_msg;
+        io_jsoncan_copyFromCanMsg(&rx_msg, &jsoncan_rx_msg);
+        io_canRx_updateRxTableWithMessage(&jsoncan_rx_msg);
+    }
 }
 
 void tasks_run1kHz(void)
 {
-    // TODO: Setup tasks.
-    osDelay(osWaitForever);
+    static const TickType_t period_ms   = 1U;
+    static uint32_t         start_ticks = 0;
+    start_ticks                         = osKernelGetTickCount();
+
+    for (;;)
+    {
+        const uint32_t start_time_ms = osKernelGetTickCount();
+
+        const uint32_t task_start_ms = TICK_TO_MS(osKernelGetTickCount());
+        io_canTx_enqueueOtherPeriodicMsgs(task_start_ms);
+
+        start_ticks += period_ms;
+        osDelayUntil(start_ticks);
+    }
 }
 
 void tasks_run1Hz(void)
 {
-    // TODO: Setup tasks.
-    osDelay(osWaitForever);
+    static const TickType_t period_ms   = 1000U;
+    static uint32_t         start_ticks = 0;
+    start_ticks                         = osKernelGetTickCount();
+
+    for (;;)
+    {
+        hw_stackWaterMarkConfig_check();
+        app_stateMachine_tick1Hz();
+
+        const bool debug_mode_enabled = app_canRx_Debug_EnableDebugMode_get();
+        io_canTx_enableMode(CAN_MODE_DEBUG, debug_mode_enabled);
+        io_canTx_enqueue1HzMsgs();
+
+        start_ticks += period_ms;
+        osDelayUntil(start_ticks);
+    }
 }
 
```

Comment:
you need to call `io_coolant_inputCaptureCallback` in a function called `void HAL_TIM_IC_CaptureCallback(TIM_HandleTypeDef *htim)`